File: 2600-maximum-tastiness-of-candy-basket/2600-maximum-tastiness-of-candy-basket.py
from functools import cache
from itertools import combinations
from typing import Iterable
from collections import Counter

class Solution:
    def maximumTastiness(self, price: List[int], k: int) -> int:
        n = len(price)
        price.sort()

        def achievable(target_tastiness):
            indices = [0]
            for i in range(1, n):
                if price[i] - price[indices[-1]] < target_tastiness:
                    continue
                indices.append(i)
                if len(indices) >= k:
                    return True
            return False

        beg = 0; end = 10 ** 9 + 1

        while beg + 1 < end:
            mid = (beg + end) // 2
            if achievable(mid):
                beg, end = mid, end
            else:
                beg, end = beg, mid
        return beg

File: 2600-maximum-tastiness-of-candy-basket/2600-maximum-tastiness-of-candy-basket.java
class Solution {
    private int[] price;
    private int k;
    
    public int maximumTastiness(int[] price, int k) {
        Arrays.sort(price);
        this.price = price;
        this.k = k;
        
        return narrowDown(0, (int) (Math.pow(10, 9) + 1));
    }
    private int narrowDown(int beg, int end) {
        var sz = end - beg;
        if (sz == 1) {
            return beg;
        }
        int mid = (beg + end) / 2;
        return possible(mid) ? narrowDown(mid, end) : narrowDown(beg, mid);
    }
    
    private boolean possible(int targetTastiness) {
        int lastPrice = this.price[0];
        int answer = 1;
        for (int i = 1; i < this.price.length && answer < this.k; i++) {
            if (this.price[i] - lastPrice < targetTastiness)
                continue;
            lastPrice = this.price[i];
            answer += 1;
        }
        return answer >= this.k;
    } 
}